### ISSUE TITLE
Upgrade to CEF 3112

### ISSUE THREAD

**Issue by Xaymar** [obsproject/obs-browser#86]
Browser Source in OBS currently is still on branch 2987, while Chrome and CEF both recently made branch 3112 "stable". It works much better, seems to have a lower CPU impact and 60 fps (with and without --enable-gpu) actually works. All that is needed is an up-to-date CEF and PR #81 applied to Browser Source.

Examples of 60 fps output while gaming:

- without --enable-gpu: [2017-08-07_02-35-19-x264.zip](https://github.com/kc5nra/obs-browser/files/1203622/2017-08-07_02-35-19-x264.zip)
- with --enable-gpu: [2017-08-07_02-53-50-x264.zip](https://github.com/kc5nra/obs-browser/files/1203623/2017-08-07_02-53-50-x264.zip)

**Comment by Lange** [comment]
I've just confirmed that a graphic I'm working on crashes CEF 2987 100% of the time. This same crash does not happen on 3112.

Here's a screenshot of the error in question, taken from my tests with the CEF sample application that's available from the [Spotify builds](http://opensource.spotify.com/cefbuilds/index.html):
![image](https://user-images.githubusercontent.com/873012/29394733-ee72df5e-82d1-11e7-98c3-bcc15979f9c4.png)

**Comment by OsirisNL** [comment]
I've uploaded a version of the plugin built with CEF 3112: https://github.com/kc5nra/obs-browser/releases/download/1.30/obs-browser-1.30.-.3112.zip

Will probably make it into OBS Studio next patch.

**Comment by Lange** [comment]
@OsirisNL just noticed that I'm having significant performance regressions with the 3112 build. Any chance you could make a 3163 build that I could try (or whatever the latest build is when you can get around to it)?

**Comment by Xaymar** [comment]
There seems to be a big issue with CEF versions newer than 2987: They leak memory like a madman. And this is a continous leak, not an instant one. You can easily trigger this leak with the Streamlabs Tip/Bit Jar and then watch the memory continously grow forever. The same steps in 2987 work just fine and also use less CPU.

**Comment by Xaymar** [comment]
Turns out that the issue is from compiling the CEF created process using the dynamic MSVC runtime (/MD) instead of the libcef runtime (/MT). This somehow led to the main process forgetting to release memory and doubled CPU usage of that process.

1. lib_cef_wrapper and plugin must match MSVC runtime (/MD or /MT)
2. cef-bootstrap and libcef must match MSVC runtime (/MT)

**Comment by Xaymar** [comment]
Turns out that the above is also wrong. It seems that CEF >= 3071 added something that obs-browser is not doing. From here on we have two options:

1. Figure out what and fix obs-browser.
2. Rewrite obs-browser from scratch and do it correctly.

**Comment by Lange** [comment]
For a short-term solution, I would be interested in trying a build based on CEF `3.3029.1619`, and seeing if it retains the performance of 2987 while avoiding the crashes I ran into.

**Comment by Lange** [comment]
It may also be worth comparing `obs-browser` implement to the in found in https://github.com/CasparCG/Server/pull/611.

Of note:
1) Most if my graphics perform better on that build of Caspar than they do in OBS.
2) However, graphics which include WebM videos with alpha channels perform far worse in that build of Caspar than they do in OBS.

In short: this build of Caspar doesn't necessarily have _better_ performance, but rather _different_ performance, with some graphics performing considerably better compared to OBS, but others performing worse.

**Comment by Lange** [comment]
I made a build with CEF `3.3029.1619`, and performance is indeed much better, seemingly similar to what it was with the 2987 build. Additionally, 3029 does not have the crashes that I experienced with 2987. I recommend that the next release be based on 3029 until we can figure out what changed in 3071.

**Comment by Lange** [comment]
I've done more research with the help of RytoEX in IRC, and we're thinking that this is an issue in CEF itself, and not necessarily an issue with `obs-browser`. Here's the data:

1) The latest available build of 3029 (CEF 3.3029.1619.geeeb5d7) performs quite well. Not perfect, but its very similar (if not identical) to what we were used to with 2987. The CEF sample client (with OSR enabled) also performs well.
2) The very next build available from Spotify is CEF 3.3071.1634.g9cc59c8, which has the very poor performance that we've been seeing. **The CEF sample client (with OSR enabled) also performs just as poorly**.
3) The latest available build (CEF 3.3163.1671.g700dc25) seems to have slightly better performance than 3071, but still much worse than 3029.

In summary, the issue does seem to be within CEF itself, as these performance issues are also present in the sample applications when they have OSR enabled. At this time, I have no idea what could have changed in 3071 to introduce such a dramatic reduction in performance.

**Comment by Xaymar** [comment]
I can't confirm point 2. For me the test applications (tests/cefclient) performed on par with Google Chrome in all situations - OSR was enabled for this test.

**Comment by RytoEX** [comment]
I think it would be helpful to establish the steps needed to compare performance, both between obs-browser and cefclient and between versions of cefclient itself.  It would be especially useful to use the same graphics or test cases.  The crashes in 2987 were, most likely, actual bugs.  The performance issues in 3071+ seem to be regressions.

@OsirisNL has been using this to test performance:  http://www.kevs3d.co.uk/dev/canvasmark/
If there are other tools that can be used to reliably test performance, please feel free to mention them.

As far as I know, the flags for cefclient to get it to run as similarly to obs-browser as possible are:

`cefclient.exe --off-screen-rendering-enabled --disable-gpu --disable-gpu-compositing`

You'll probably want to specify `--off-screen-frame-rate=60`, otherwise it'll default to 30 FPS.

For CEF < 3017, you'll also need `--transparent-painting-enabled`.  I _think_ that cefclient will also append the switch `--disable-surfaces` [by default if you enable OSR](https://github.com/chromiumembedded/cef/blob/3071/tests/shared/browser/client_app_browser.cc#L24).  Though, that's mostly for rendering PDFs, so I don't know if that matters for obs-browser or testing between them.

It seems that obs-browser currently uses `--enable-begin-frame-scheduling`, though cefclient removed that in CEF branch 2840 in [this commit](https://bitbucket.org/chromiumembedded/cef/commits/9064e302961ea11134aebcaa33d53efa538e8430?at=2840).  Unfortunately, that commit message makes no mention of this or why, so I can only suspect that an underlying change in Chromium [between fc6aaca4 and 1ae106db](https://chromium.googlesource.com/chromium/src/+log/fc6aaca4..1ae106db) (or farther back) made it unnecessary?  CefSharp set it as on by default [after CEF added it](https://bitbucket.org/chromiumembedded/cef/commits/e3c1d8632eb43c1c2793d71639f3f5695696a5e8#Ltests/cefclient/browser/client_app_browser.ccT47), but didn't remove it after CEF did.

If anything above is incorrect, or there's a better set of parameters or a better way to test, please feel free to say so and further contribute to the discussion.  I hate to suggest bisecting CEF between 2987 and 3071, because building CEF is painful, but that may have to be an option, especially if this persists in CEF 3202.

-----

I've been reading up on `--disable-gpu-vsync`, and apparently there have been issues, and patches, and more issues.  Some issues occur with vsync enabled and some with it disabled.  Some of the issues are/were in ANGLE.  I hope the issues seen in this thread aren't as deep as those.
